pkg/util/pravegacluster.go

```diff
@@ -226,3 +226,77 @@ func NormalizeVersion(version string) (string, error) {
 	}
 	return matches[1], nil
 }
+
+type OrderedMap struct {
+	m    map[string]string
+	keys []string
+}
+
+func UpdateOneJVMOption(arg string, om *OrderedMap) {
+	if strings.HasPrefix(arg, "-Xms") {
+		if _, ok := om.m["-Xms"]; !ok {
+			om.keys = append(om.keys, "-Xms")
+		}
+		om.m["-Xms"] = arg[4:]
+		return
+	}
+
+	if strings.HasPrefix(arg, "-XX:") {
+		if arg[4] == '+' || arg[4] == '-' {
+			if _, ok := om.m[arg[5:]]; !ok {
+				om.keys = append(om.keys, arg[5:])
+			}
+			om.m[arg[5:]] = string(arg[4])
+			return
+		}
+		s := strings.Split(arg[4:], "=")
+		if _, ok := om.m[s[0]]; !ok {
+			om.keys = append(om.keys, s[0])
+		}
+		om.m[s[0]] = s[1]
+		return
+	}
+
+	if _, ok := om.m[arg]; !ok {
+		om.keys = append(om.keys, arg)
+	}
+	om.m[arg] = ""
+	return
+}
+
+func GenerateJVMOption(k, v string) string {
+	if v == "" {
+		return k
+	}
+
+	if k == "-Xms" {
+		return fmt.Sprintf("%v%v", k, v)
+	}
+
+	if v == "+" || v == "-" {
+		return fmt.Sprintf("-XX:%v%v", v, k)
+	}
+
+	return fmt.Sprintf("-XX:%v=%v", k, v)
+}
+
+func OverrideDefaultJVMOptions(defaultOpts []string, overrideOpts []string) []string {
```

Comment:
Thanks for the comment. I see your point and here is what I think,
1. If the number of default options is `m` and the number of user-provided options is `n`, this line `if (key is a prefix for any value in overrideSlice)` will basically make the complexity to be `m*n`, whereas my current way is `m+n`. 
2. Since iteration for map in Go does not follow the insertion order, this line `for key, value := range defaultOptsMap` will make the iteration in random order. However, I found that we have to specify `UnlockExperimentalVMOptions` before `UseCGroupMemoryLimitForHeap` when passing to the JVM, otherwise JVM will throw an error. So there is an order that needs to be guaranteed when passing those JVM options.

Reply:
I don't think we need to worry so much about execution complexity here, unless we expect the number of options ( default or user provided) to be > 100. In most cases the number of options provided ( default + user provided) should be not more than 10-15 ( At max this can go upto 30-35) which is too small a number to worry about execution complexity.
It is more important to have clear and readable code that is easier to maintain and reason about.
As regards the order of options, only a few options should be order sensitive and it should be easy to just add them first.

Reply:
I've added some comments to make it more readable. It would be great if you could take a look at it again and see if it works. Thanks!